src/pyodide/internal/asgi.py

```diff
@@ -98,54 +98,123 @@ async def send(got):
-            # body is corrupted. This is fine because the session is going to
-            # exit at this point.
-            px = create_proxy(got["body"])
+            # Check if this is a server-sent events response
+            for k, v in headers:
+                if k.lower() == "content-type" and v.lower().startswith(
+                    "text/event-stream"
+                ):
+                    is_sse = True
+                    break
+            if is_sse:
+                # For SSE, create and return the response immediately after http.response.start
+                resp = Response.new(
+                    readable, headers=Object.fromEntries(headers), status=status
+                )
+                result.set_result(resp)
+
+        elif got["type"] == "http.response.body":
+            body = got["body"]
+            more_body = got.get("more_body", False)
+
+            # Convert body to JS buffer
+            px = create_proxy(body)
             buf = px.getBuffer()
             px.destroy()
-            resp = Response.new(
-                buf.data, headers=Object.fromEntries(headers), status=status
-            )
-            result.set_result(resp)
 
-    await app(request_to_scope(req, env), receive, send)
-    return await result
+            if is_sse:
+                # For SSE, write chunk to the stream
+                await writer.write(buf.data)
+                # If this is the last chunk, close the writer
+                if not more_body:
+                    await writer.close()
+                    finished_response.set()
+            else:
+                resp = Response.new(
+                    buf.data, headers=Object.fromEntries(headers), status=status
+                )
+                result.set_result(resp)
+                await writer.close()
+                finished_response.set()
+
+    # Run the application in the background to handle SSE
+    async def run_app():
+        try:
+            await app(request_to_scope(req, env), receive, send)
+
+            # If we get here and no response has been set yet, the app didn't generate a response
+            if not result.done():
+                raise RuntimeError("The application did not generate a response")  # noqa: TRY301
+        except Exception as e:
+            # Handle any errors in the application
+            if not result.done():
+                result.set_exception(e)
+                await writer.close()  # Close the writer
+                finished_response.set()
+
+    # Create task to run the application in the background
+    app_task = create_task(run_app())
+
+    # Wait for the result (the response)
+    response = await result
+
+    # For non-SSE responses, we need to wait for the application to complete
+    if not is_sse:
+        await app_task
+    else:  # noqa: PLR5501
+        if ctx is not None:
+            ctx.waitUntil(create_once_callable(app_task))
```

Comment:
Adding a comment to block this PR until @hoodmane approves this line.

Reply:
This seems odd don't you `waitUntil` a promise and not a callable?

Reply:
I pinged you on the internal chat
There was an error trying to wait until a Python promise
I thought this would convert it to a JavaScript promise 
It seemed to work
The blocking comment is exactly because I have no honest clue how to convert a Python promise to a JavaScript promise but this seemed to work

Reply:
I think this should be `create_proxy()` here and not `create_once_callable()`. Otherwise it looks good enough for now.

Reply:
How would I use create_proxy in this scenario?

Reply:
FWIW create_once_callable is working well both here and deployed
Why is it not appropriate
What's the difference?
The documentation is very very scarce

Reply:
Well `create_once_callable` is generally only for callables. Promises are not callables. I would think that `ctx.waitUntil(create_once_callable(p))` would be roughly a no-op, same as if you hadn't called `ctx.waitUntil()` at all, because the callable is not a thenable, and so when it awaits the non-thenable it resolves immediately to the callable.

`create_once_callable()` should type check the argument to ensure that it is a callable and throw in cases like this where it isn't. Anyways `ctx.waitUntil(create_proxy(app_task))` should do what you want.

Reply:
Updated